src/query/service/src/sql/executor/pipeline_builder.rs

```diff
@@ -472,4 +481,43 @@ impl PipelineBuilder {
         // ExchangeSink will be appended by `ExchangeManager::execute_pipeline`
         self.build_pipeline(&exchange_sink.input)
     }
+
+    fn expand_union(&mut self, plan: &PhysicalPlan) -> Result<Receiver<Option<DataBlock>>> {
+        let union_ctx = QueryContext::create_from(self.ctx.clone());
+        let pipeline_builder = PipelineBuilder::create(union_ctx);
+        let mut build_res = pipeline_builder.finalize(plan)?;
+
+        assert!(build_res.main_pipeline.is_pulling_pipeline()?);
+
+        build_res.main_pipeline.resize(1)?;
+        let (tx, rx) = channel(1);
+        let input = InputPort::create();
+        build_res.main_pipeline.add_pipe(Pipe::SimplePipe {
+            outputs_port: vec![],
+            inputs_port: vec![input.clone()],
+            processors: vec![UnionReceiveSink::create(tx, input)],
+        });
+        self.pipelines.push(build_res.main_pipeline);
+        self.pipelines
+            .extend(build_res.sources_pipelines.into_iter());
+        Ok(rx)
+    }
+
+    pub fn build_union(&mut self, union: &Union) -> Result<()> {
+        self.build_pipeline(&union.left)?;
+        let union_receiver = self.expand_union(&union.right)?;
+        self.main_pipeline.resize(1)?;
```

Comment:
Not need this if we use async_channel::bounded(2)?

Reply:
Yes, fixed